AEP-9936: Per-VPA Observation Window (#9962)
kubernetes/autoscaler: vertical-pod-autoscaler/enhancements/9936-vpa-observation-window/README.md

  - [Validation](#validation)
  - [Status Condition](#status-condition)
  - [Metric](#metric)
  - [Feature Enablement and Rollback](#feature-enablement-and-rollback)
  - [Kubernetes Version Compatibility](#kubernetes-version-compatibility)
- [Test Plan](#test-plan)
- [Examples](#examples)
  - [Basic: <code>Recreate</code> with a six-hour window](#basic-recreate-with-a-six-hour-window)
  - [Fast iteration: <code>InPlace</code> with a short window](#fast-iteration-inplace-with-a-short-window)
  - [Pod-creation gating: <code>Initial</code> with a one-hour window](#pod-creation-gating-initial-with-a-one-hour-window)
- [Alternatives Considered](#alternatives-considered)
- [Implementation History](#implementation-history)
<!-- /toc -->

## Summary

Add an optional `observationPeriodSeconds` field to `PodUpdatePolicy` that tells the Updater to refrain from applying recommendations for a configurable duration after a VPA's creation. During the window the Recommender continues to compute and publish recommendations to `status.recommendation` normally — only the Updater is gated. Once the window elapses, the configured `updateMode` takes effect automatically with no operator action.

> The window is computed as a pure function of `vpa.CreationTimestamp` and the current spec value, evaluated on every reconcile. No status writes are required.

Reviewer: This needs further elaboration.
For example, if I modify an existing VPA's spec, does the `observationPeriodSeconds` window reset?
After reading the doc I see that you have explained it below. might be worth to also put it here.

Author: Done — the Summary now states this directly: modifying the spec never resets the window, since the anchor is the immutable CreationTimestamp; changing initialDelaySeconds only moves the expiry relative to it. Full PATCH semantics are in the Gate Evaluation table.

When new `Deployment` objects are provisioned automatically — by CI, per-tenant provisioning, on-demand application creation, etc. — a VPA created alongside them and configured with a non-`Off` `updateMode` will start actuating on recommendations that have not stabilised yet. Recommendations are especially unstable in the first few hours of a workload's life because the recommender has few usage samples, and applying those recommendations causes disruptive churn: pods are restarted with the wrong resource requests, then restarted again a few minutes later with different requests.

The safe pattern is to run the VPA in `Off` mode for the first N hours to let recommendations stabilise, then flip to `Recreate` / `InPlaceOrRecreate` / `InPlace` / `Initial`. Doing that manually for every new VPA does not scale: cluster operators end up building external controllers, CronJobs, or admission-side patch scripts to sequence the mode transition. Every environment reinvents the same workflow.

This feature moves the sequencing into the VPA API itself. Operators declare "wait N seconds after creation before touching pods" alongside the update mode, and the Updater honours it.

### Goals

> - Provide a declarative per-VPA field that delays Updater actuation for a configurable duration after VPA creation.

Reviewer 2: Should admission-controller also be in scope here?
What happens if a new pod is created during this period? It will get the wrong recommendation

Author: Good catch — yes, it has to be. Without gating admission, any pod created during the window would be injected with exactly the un-stabilised recommendations the window is meant to hold back. Updated the AEP: the Admission Controller now evaluates the same gate and admits pods with their original spec resources while the window is active

- Keep the Recommender's behaviour unchanged: recommendations are still computed and published to `status.recommendation` during the window, so operators can inspect them.

> - Work uniformly across `Recreate`, `InPlaceOrRecreate`, `InPlace`, and `Initial`.

Reviewer 2: How does this feature interact with CPUStartupBoost?

Reviewer: I believe that once `CPUStartupBoost` is over and the updater needs to perform an in-place update of the pod, it should follow the same logic. if the startup boost window has not yet passed, it should scale to the resource spec; otherwise, it should scale to the VPA recommendations.

Reviewer 2: :)

It's a leading question, the answer needs to be documented in the AEP

Reviewer: Agree.

Author: Sure, Documented in a new "Interaction with CPU Startup Boost" section.

- Surface the gate state through both a status condition (`ObservationPeriodActive`) and a Prometheus gauge (`vpa_in_observation_window`), so operators and dashboards can see when a VPA is gated without inspecting `spec`.

### Non-Goals

- **Recommender-confidence-based gating.** This feature is a user-declared policy, not a recommender-computed signal. Interoperation with `LowConfidence` / `RecommendationProvided` is left for future work.
- **Reset semantics on PATCH of other fields.** Changing `resourcePolicy`, `targetRef`, or the target workload does not reset the window. The window is anchored on `vpa.CreationTimestamp`; users who need to re-observe a workload after a material change should delete and recreate the VPA.
- **Cluster-wide default windows.** Different applications and different VPAs on the same target want different windows; this is a per-VPA concern.

## Proposal

Add a single optional field to `PodUpdatePolicy` (autoscaling.k8s.io/v1):

```go
// ObservationPeriodSeconds is the duration in seconds after this VPA's
// creation during which the Updater will refrain from applying
// recommendations, regardless of the configured UpdateMode.
// +optional

> // +kubebuilder:validation:Minimum=1

Reviewer 2: Can this field also have a maximum?
See https://github.com/kubernetes/community/blob/e9959eca8c650dd9ad9dc2809954fc1984c6d9b4/contributors/devel/sig-architecture/api-conventions.md#numeric-fields

Author: Added — +kubebuilder:validation:Maximum=7776000 (90 days).

> ObservationPeriodSeconds *int32 `json:"observationPeriodSeconds,omitempty"`

Reviewer 2: I'm not sure I'm excited by this name. 

What about `InitialDelaySeconds`? (Matching what Pod probes call it)

Reviewer 2: `ActuationDelaySeconds` ?

Reviewer: `InitialDelaySeconds` . I like that.

Author: Sure, `initialDelaySeconds` sounds better. Renamed to `initialDelaySeconds` throughout

## Design Details

### Workflow

1. The user creates a VPA with a non-`Off` `updateMode` and `observationPeriodSeconds: N` set on `spec.updatePolicy`.
2. The Recommender begins computing recommendations on the normal schedule and populates `status.recommendation`. It is unaware of and unaffected by the observation window.
3. On every Updater reconcile, before deciding whether the VPA is eligible for actuation, the Updater evaluates the gate. If the gate is active, the VPA is treated as if `updateMode` were `Off` for that reconcile: no pods are evicted (`Recreate` / `InPlaceOrRecreate`), no in-place resize is attempted (`InPlace` / `InPlaceOrRecreate`), and no admission-controller-side injection happens on new-pod creation (`Initial`).
4. The Updater sets the `ObservationPeriodActive` status condition to `True` while the gate is active and `False` once it has elapsed. It emits the `vpa_in_observation_window` gauge accordingly.
5. On the first reconcile after `CreationTimestamp + observationPeriodSeconds`, the gate opens and the configured `updateMode` takes effect on subsequent reconciles.

> ### API Changes

Reviewer: We will probably need to add a condition for that too.
After reading the doc I see that you have explained it below. might be worth to link it here.

Author: Linked — the Proposal section now points to the InitialDelayActive condition where the gate is first described.

// applying recommendations.
func inObservationWindow(vpa *vpa_types.VerticalPodAutoscaler, now time.Time) bool {
    p := vpa.Spec.UpdatePolicy
    if p == nil || p.ObservationPeriodSeconds == nil || *p.ObservationPeriodSeconds < 1 {
        return false
    }
    expiry := vpa.CreationTimestamp.Add(
        time.Duration(*p.ObservationPeriodSeconds) * time.Second,
    )
    return now.Before(expiry)
}
```

Insertion point: at the top of the Updater's eligibility check in `pkg/updater/logic/updater.go`, at the same call site where `UpdateModeOff` and `UpdateModeInitial` currently short-circuit the eviction path. When `inObservationWindow` returns `true`, the Updater takes the same code path it takes for `UpdateModeOff` today.

The gate is stateless: it is a pure function of `CreationTimestamp` (immutable on the object) and the current spec value (mutable). No caching, no status writes.

Consequences that fall out of this design:

| PATCH operation | Effect |
| --- | --- |
| Shorten `observationPeriodSeconds` during window | Gate releases earlier on next reconcile. |
| Extend `observationPeriodSeconds` during window | Gate stays open longer. |

> | Extend `observationPeriodSeconds` **after** expiry | Gate re-arms on next reconcile. |

Reviewer 2: Is a re-arm here guaranteed? 
Doesn't it only get re-armed if the new `observationPeriodSeconds` hasn't elapsed (when compared to creation time) ?

Author: > Is a re-arm here guaranteed?

No, not guaranteed

> Doesn't it only get re-armed if the new observationPeriodSeconds hasn't elapsed (when compared to creation time) ?

Yes, correct
I have clarified this properly

| Remove `observationPeriodSeconds` (or set to zero) | Gate closes immediately on next reconcile. |
| PATCH `updateMode` | New mode takes effect when the window elapses (or immediately, if already elapsed). |

> | PATCH `resourcePolicy` / `targetRef` | No effect on the gate. |

Reviewer: So, does that mean that patches to other fields in the VPA spec don't affect this "gate" at all?

Author: Yes, they don't. My thought process was to keep it simple and intuitive for users. Though if you feel differently, we can revisit, I am open to suggestions

> | PATCH `resourcePolicy` / `targetRef` | No effect on the gate. |

Reviewer 2: ```suggestion
| Modify `updateMode` | New mode takes effect when the window elapses (or immediately, if already elapsed). |
| Modify `resourcePolicy` / `targetRef` | No effect on the gate. |
```

The re-arm-after-expiry behaviour is intentional: it lets users manually extend an observation window in response to observed instability without deleting the VPA. If reviewers consider this a footgun serious enough to justify admission logic, we can add a check rejecting PATCHes that would push `expiry` past `now` once the window has already elapsed — see [Alternatives Considered](#alternatives-considered).

### Interaction with `updateMode`

The observation window is orthogonal to `updateMode`. While the gate is active, the Updater behaves as if `updateMode` were `Off` regardless of the configured value. Once the gate opens, the configured mode takes effect normally:

- `Recreate` → the Updater is now eligible to evict pods to apply recommendations.
- `InPlaceOrRecreate` → the Updater attempts in-place resize first, falling back to eviction.
- `InPlace` → the Updater attempts in-place resize only; no evictions.
- `Initial` → the admission-controller-side injection is now eligible to run on subsequent new-pod creations. Pods created during the window (or pods that pre-dated the VPA) are unaffected, matching the standard `Initial` semantics.

> - `Off` → `observationPeriodSeconds` has no effect. `Off` already suppresses actuation indefinitely, so combining the two is a valid (but redundant) configuration that silently no-ops. No error, no warning; the field is simply ignored.

Reviewer 2: This detail seems unnecessary.
This listing already defined behaviour

Author: Applied 👍

### Validation

Enforced via CRD schema:

- `observationPeriodSeconds` must be an integer `>= 1` if set.

No dynamic (admission-webhook) validation is required in v1. All lifecycle transitions are handled declaratively by the gate.

### Status Condition

Add a new value to the existing `VerticalPodAutoscalerStatus.Conditions` slice:

- **Type:** `ObservationPeriodActive`
- **Status:** `True` while the gate is active, `False` once it has elapsed.
- **Reason:** `WindowActive` (True) / `WindowExpired` (False).
- **Message:** human-readable summary including expiry timestamp.

This lets `kubectl describe vpa` surface the gate without an operator having to compute `CreationTimestamp + observationPeriodSeconds` mentally.

### Metric

Emit a new gauge from the Updater:

```
vpa_in_observation_window{namespace, name} = 0 | 1
```

> Value is `1` while the gate is active for that VPA, `0` otherwise. Enables fleet-wide visibility into how many VPAs are currently gated, useful for dashboards and alerting on stuck windows.

Reviewer: Does that really useful? users can create alerts using the condition state. no strong opinions here.

Author: In my opinion, its useful. For example, for our usecase, we have 100s of VPA objects over 100s of different vClusters, so its useful for us to see it in dashboards, and build alerts on it

Reviewer: I'd still not convinced this metric is needed if you have a condition for it. but maybe it's just me.

**E2E tests** (`e2e/`):

- Create a VPA with `updateMode: Recreate` and `observationPeriodSeconds: 60`. Verify no eviction occurs during the first 60 s. Advance time past the window; verify the next reconcile evicts as expected.
- Repeat for `InPlaceOrRecreate`, `InPlace`, and `Initial` (with pod-creation-timing tests for the last).
- Create a VPA with `observationPeriodSeconds: 3600`, PATCH to `60` after 30 s, verify eviction becomes eligible around the 60 s mark.
- Create a VPA with `observationPeriodSeconds: 60`, wait for expiry, PATCH to `3600`, verify the gate re-arms and eviction is again suppressed until the new expiry.
- Verify the `ObservationPeriodActive` condition and `vpa_in_observation_window` gauge match the gate state throughout each scenario.

## Examples

### Basic: `Recreate` with a six-hour window

Typical "let this new service settle for a work-day" configuration.

```yaml
apiVersion: autoscaling.k8s.io/v1
kind: VerticalPodAutoscaler
metadata:
  name: web-vpa
  namespace: production
spec:
  targetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: web
  updatePolicy:
    updateMode: Recreate
    observationPeriodSeconds: 21600   # 6 hours
```

> For the first six hours after creation, `web` pods are not evicted regardless of recommendations. After the window, `Recreate` behaviour takes effect.

Reviewer 2: ```suggestion
For the first six hours after creation, `web` pods are not evicted by the updater regardless of recommendations. After the window, `Recreate` behaviour takes effect.
```

  - [Kubernetes Version Compatibility](#kubernetes-version-compatibility)
- [Test Plan](#test-plan)
- [Examples](#examples)
  - [Basic: <code>Recreate</code> with a six-hour window](#basic-recreate-with-a-six-hour-window)
  - [Fast iteration: <code>InPlace</code> with a short window](#fast-iteration-inplace-with-a-short-window)
  - [Pod-creation gating: <code>Initial</code> with a one-hour window](#pod-creation-gating-initial-with-a-one-hour-window)
- [Alternatives](#alternatives)
- [Implementation History](#implementation-history)
<!-- /toc -->

## Summary

Add an optional `initialDelaySeconds` field to `PodUpdatePolicy` that delays actuation of recommendations for a configurable duration after a VPA's creation. During the window the Recommender continues to compute and publish recommendations to `status.recommendation` normally, but nothing actuates them: the Updater performs no evictions or in-place resizes, and the Admission Controller does not apply recommendations to pods created during the window. The VPA behaves exactly as if `updateMode` were `Off`. Once the window elapses, the configured `updateMode` takes effect automatically with no operator action.

> The window is computed as a pure function of `vpa.CreationTimestamp` and the current spec value, evaluated on every reconcile. Modifying an existing VPA's spec does not reset the window — the anchor (`CreationTimestamp`) is immutable, and changing `initialDelaySeconds` itself simply moves the expiry relative to that anchor (see [Gate Evaluation](#gate-evaluation) for the full modification semantics).

Reviewer: What does "and the current spec value" mean here? later you are saying that  "Modifying an existing VPA's spec does not reset the window".

Author: By current spec value, I meant the current `initialDelaySeconds` value. 

"The window is computed as a pure function of `vpa.CreationTimestamp` and the current spec value, evaluated on every reconcile." - I mean to say that the window is created from the creation timestamp + the `initialDelaySeconds`

"Modifying an existing VPA's spec does not reset the window" - Since the creation timestamp and the initalDelaySeconds hasn't changed, the observation window won't be updated.

Made a minor adjustment to make it clearer - 

"current spec value" → "current `initialDelaySeconds` value."

    }
    expiry := vpa.CreationTimestamp.Add(
        time.Duration(*p.InitialDelaySeconds) * time.Second,
    )
    return now.Before(expiry)
}
```

Insertion points:

- **Updater** — at the top of the eligibility check in `pkg/updater/logic/updater.go`, at the same call site where `UpdateModeOff` and `UpdateModeInitial` currently short-circuit the eviction path.
- **Admission Controller** — at the existing `UpdateModeOff` short-circuits in `pkg/admission-controller/resource/vpa/matcher.go` and `pkg/admission-controller/resource/pod/recommendation/recommendation_provider.go`, so pods created during the window are admitted with their original spec resources.

When `InInitialDelayWindow` returns `true`, both components take the same code path they take for `UpdateModeOff` today.

`InInitialDelayWindow` is exported from the shared `pkg/utils/vpa` package so both the Updater and the Admission Controller consume the identical implementation. Each call site first checks the `VPAInitialDelay` feature gate (`features.Enabled(features.VPAInitialDelay)`); when the gate is disabled the helper is not consulted and the component behaves exactly as it does today (fail-open).

> The gate is stateless: it is a pure function of `CreationTimestamp` (immutable on the object) and the current spec value (mutable). No caching, no status writes.

Reviewer: Same as above.

### Non-Goals

- **Recommender-confidence-based gating.** This feature is a user-declared policy, not a recommender-computed signal.
- **Reset semantics on modification of other fields.** Changing `resourcePolicy`, `targetRef`, or the target workload does not reset the window. The window is anchored on `vpa.CreationTimestamp`; users who need to re-observe a workload after a material change should delete and recreate the VPA.
- **Cluster-wide default windows.** Out of scope for this AEP; operators can inject a default with a Mutating Admission Policy (see [Alternatives](#alternatives)).

## Proposal

Add a single optional field to `PodUpdatePolicy` (autoscaling.k8s.io/v1):

```go
// InitialDelaySeconds is the duration in seconds after this VPA's
// creation during which recommendations are not actuated: the Updater
// performs no evictions or in-place resizes, and the Admission
// Controller does not apply recommendations to newly created pods,
// regardless of the configured UpdateMode.
// +optional
// +kubebuilder:validation:Minimum=1
// +kubebuilder:validation:Maximum=7776000
InitialDelaySeconds *int32 `json:"initialDelaySeconds,omitempty"`
```

> Behaviour, in one sentence: **the Updater and the Admission Controller treat the VPA as if `updateMode` were `Off` until `now >= vpa.CreationTimestamp + spec.updatePolicy.initialDelaySeconds`.** After that, the configured `updateMode` takes effect.

Reviewer: I don't like the excessive wording. let's just write here something like: "With the new InitialDelaySeconds the updater and the admission controller treat the VPA as if updateMode were Off until `now >= vpa.CreationTimestamp + spec.updatePolicy.initialDelaySeconds`"

### Interaction with CPU Startup Boost

CPU Startup Boost (AEP-7862) still applies during the delay window: the boost is a startup-safety mechanism computed from the pod's own spec, not a recommendation, and the Admission Controller already applies it even under `updateMode: Off` (the `Off` short-circuit in `resource/vpa/matcher.go` is bypassed for VPAs with a boost). The gate reuses the same `Off` semantics, so boosted containers remain boosted.

The interaction appears at boost expiry, when the Updater performs the post-boost in-place resize:

- Delay window still active → the pod is scaled down to its original spec resources; no recommendation is actuated.
- Delay window elapsed → the pod is resized to the VPA recommendation, per normal post-boost behaviour.

### Interaction with Checkpoints

A VPA's checkpoint stores the recommender's history for that VPA. Checkpoints are keyed by VPA name, not by a unique object ID, and one is only deleted once no VPA with that name exists (the recommender's GC pass, every 10 minutes by default via `--checkpoints-gc-interval`).

> So deleting a VPA and recreating it with the same name hands the new object the old checkpoint, as long as GC hasn't removed it yet. The recommender loads that history and fills `status.recommendation` with mature values almost immediately, even though the new VPA has a fresh UID and creation time.

Reviewer: So in other words - checkpoint are not affected by the new feature (and will still be populated with values), and the observer gate depends only on creation timestamp?

Author: Yes, that's correct.

```yaml
apiVersion: autoscaling.k8s.io/v1
kind: VerticalPodAutoscaler
metadata:
  name: tenant-worker-vpa
  namespace: tenant-42
spec:
  targetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: worker
  updatePolicy:
    updateMode: Initial
    initialDelaySeconds: 3600    # 1 hour
```

Pods created during the first hour receive their `Deployment`-spec resources unchanged. Pods created after the window get VPA-injected resources.

## Alternatives

**1. External tooling (today's status quo).** Without this feature, you get the same behavior by creating the VPA in `Off` mode and flipping it to an active mode after N hours — a CronJob or a small controller that patches `updateMode`. It works, but every team rebuilds the same automation, and there's no per-VPA declarative way to say "wait N hours before acting."

**2. Extend `EvictionRequirements` with a time-based condition.** `EvictionRequirements` (AEP-4831) already gates eviction on conditions. A duration doesn't fit its enum, which compares the recommendation against requests (`TargetHigherThanRequests` / `TargetLowerThanRequests`), and it only covers eviction — not admission-time injection or in-place resize.

> **3. Default the field with a Mutating Admission Policy.** For a cluster-wide default, a [Mutating Admission Policy](https://kubernetes.io/docs/reference/access-authn-authz/mutating-admission-policy/) can set `initialDelaySeconds` on VPAs that don't specify it, using native Kubernetes with no VPA-side flag. This complements the field rather than replacing it.

Reviewer: Not sure I understand this section, how does adding `initialDelaySeconds` to your VPA using Mutating Admission Policy is an alternative to this feature in world that the feature is not exist?

Author: yeah my bad, I misunderstood the meaning of Alternatives considered, fixed now and checked if any such mistakes exist.

## Implementation History

- (issue filed) 2026-07-06 — Issue [kubernetes/autoscaler#9936](https://github.com/kubernetes/autoscaler/issues/9936) filed.
- (triage accepted) 2026-07-07 — `/triage accepted` from SIG member.
- (AEP PR opened) 2026-07-10 — PR [kubernetes/autoscaler#9962](https://github.com/kubernetes/autoscaler/pull/9962).
- (initial implementation) TBD.
- (alpha) TBD.

> - (beta) TBD.

Reviewer: Let's replace this with the current date of this PR being created as initial version. see example here https://github.com/kubernetes/autoscaler/tree/master/vertical-pod-autoscaler/enhancements/7862-cpu-startup-boost.

Author: Done 👍

```yaml
apiVersion: autoscaling.k8s.io/v1
kind: VerticalPodAutoscaler
metadata:
  name: tenant-worker-vpa
  namespace: tenant-42
spec:
  targetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: worker
  updatePolicy:
    updateMode: Initial
    initialDelaySeconds: 3600    # 1 hour
```

Pods created during the first hour receive their `Deployment`-spec resources unchanged. Pods created after the window get VPA-injected resources.

## Alternatives Considered

**1. Extend `EvictionRequirements` with a `TimeSinceCreationExceeds` value.** Reuses an existing extensibility hook (AEP-4831), but forces a duration to be encoded inside an enum meant for direction-of-change comparisons (`TargetHigherThanRequests` / `TargetLowerThanRequests`). Awkward and probably a design regression on that API.

**2. Recommender-side confidence gate.** Block `RecommendationProvided` from becoming `True` for N hours or until a confidence threshold is met. Cleaner conceptually but hides visibility — operators would see no recommendation at all in `status.recommendation` during the window, defeating the "collect first, inspect, then apply" workflow this feature is meant to support.

> **3. Global Updater flag** (e.g. `--default-observation-window-seconds`). Wrong axis: different applications in the same cluster want different windows, and different VPAs on the same target want different windows.

Reviewer 2: While I was initially against adding yet-another-flag to the updater, having a sane default kinda makes sense.
But, it would mean adding that default to both admission-controller and updater.

> Wrong axis: different applications in the same cluster want different windows, and different VPAs on the same target want different windows

I disagree with this premise and I think a sane default may be useful, I'm just not sure I like that it needs to be configured in 2 places.

What if we provided an example Mutating Admission Policy for users who may want to set a default?
(that could be out of scope of the AEP though)

Author: Fair point, dropped the "wrong axis" framing. I like the MAP idea, so I have added an example policy

When new `Deployment` objects are provisioned automatically — by CI, per-tenant provisioning, on-demand application creation, etc. — a VPA created alongside them and configured with a non-`Off` `updateMode` will start actuating on recommendations that have not stabilised yet. Recommendations are especially unstable in the first few hours of a workload's life because the recommender has few usage samples, and applying those recommendations causes disruptive churn: pods are restarted with the wrong resource requests, then restarted again a few minutes later with different requests.

The safe pattern is to run the VPA in `Off` mode for the first N hours to let recommendations stabilise, then flip to `Recreate` / `InPlaceOrRecreate` / `InPlace` / `Initial`. Doing that manually for every new VPA does not scale: cluster operators end up building external controllers, CronJobs, or admission-side patch scripts to sequence the mode transition. Every environment reinvents the same workflow.

This feature moves the sequencing into the VPA API itself. Operators declare "wait N seconds after creation before touching pods" alongside the update mode, and the Updater honours it.

### Goals

> - Provide a declarative per-VPA field that delays all actuation — evictions, in-place resizes, and admission-time injection into newly created pods — for a configurable duration after VPA creation.

Reviewer 2: Given that a CPU startup boost is a form of actuation, I think this needs rewording:
> Provide a declarative per-VPA field that delays all actuation

Author: Reworded - the goal now says 'delays actuation of recommendations'

- Keep the Recommender's behaviour unchanged: recommendations are still computed and published to `status.recommendation` during the window, so operators can inspect them.
- Work uniformly across `Recreate`, `InPlaceOrRecreate`, `InPlace`, and `Initial`.

> - Require no status writes and no PATCH-blocking admission validation in v1: the gate is a pure function of `CreationTimestamp` and the current spec value.

Reviewer 2: Is this a goal? Seems like it can be removed

Author: Yes, removed 👍

- Surface the gate state through both a status condition (`InitialDelayActive`) and a Prometheus gauge (`vpa_initial_delay_active`), so operators and dashboards can see when a VPA is gated without inspecting `spec`.

### Non-Goals

- **Recommender-confidence-based gating.** This feature is a user-declared policy, not a recommender-computed signal. Interoperation with `LowConfidence` / `RecommendationProvided` is left for future work.

> - **Reset semantics on PATCH of other fields.** Changing `resourcePolicy`, `targetRef`, or the target workload does not reset the window. The window is anchored on `vpa.CreationTimestamp`; users who need to re-observe a workload after a material change should delete and recreate the VPA.

Reviewer 2: ```suggestion
- **Reset semantics on modification of other fields.** Changing `resourcePolicy`, `targetRef`, or the target workload does not reset the window. The window is anchored on `vpa.CreationTimestamp`; users who need to re-observe a workload after a material change should delete and recreate the VPA.
```

(this document is for humans, not an API)

When `inInitialDelayWindow` returns `true`, both components take the same code path they take for `UpdateModeOff` today.

The gate is stateless: it is a pure function of `CreationTimestamp` (immutable on the object) and the current spec value (mutable). No caching, no status writes.

Consequences that fall out of this design:

| PATCH operation | Effect |
| --- | --- |
| Shorten `initialDelaySeconds` during window | Gate releases earlier on next reconcile. |
| Extend `initialDelaySeconds` during window | Gate stays open longer. |
| Extend `initialDelaySeconds` **after** expiry | Gate re-arms on the next reconcile **only if** the new expiry (`CreationTimestamp + newValue`) still lies in the future. A value whose expiry is already in the past leaves the gate open. |
| Remove `initialDelaySeconds` (or set to zero) | Gate closes immediately on next reconcile. |
| Modify `updateMode` | New mode takes effect when the window elapses (or immediately, if already elapsed). |

> | Modify `resourcePolicy` / `targetRef` | No effect on the gate. |

Reviewer: I'm still uncomfortable with why some fields affect the gate while others don't. 
Also, In the future if/when we extend the VPA CRD further, does that mean we'll also need to consider whether each new field should affect the gate?

Author: > I'm still uncomfortable with why some fields affect the gate while others don't.

No field affects the gate. The gate is a single formula - now < CreationTimestamp + initialDelaySeconds — with exactly two inputs, one immutable and one being the field itself. `initialDelaySeconds` changes the outcome only because it's the formula's parameter, and updateMode only changes what happens after the gate opens, not the gate itself.

Reviewer: sgtm

Reviewer: So can we change the wording? the table is a little bit confusing. I suggest to remove the table and just saying that no vpa change affect the gate **except** changing the value of `initialDelaySeconds` which can determine if the gate is open or closed based on `CreationTimestamp`

Author: Done- replaced the table

### Interaction with `updateMode`

The observation window is orthogonal to `updateMode`. While the gate is active, the Updater and the Admission Controller behave as if `updateMode` were `Off` regardless of the configured value: no evictions, no in-place resizes, and no recommendation injection into pods created during the window. Once the gate opens, the configured mode takes effect normally:

- `Recreate` → the Updater is now eligible to evict pods to apply recommendations.
- `InPlaceOrRecreate` → the Updater attempts in-place resize first, falling back to eviction.
- `InPlace` → the Updater attempts in-place resize only; no evictions.

> - `Initial` → the admission-controller-side injection is now eligible to run on subsequent new-pod creations. Pods created during the window (or pods that pre-dated the VPA) are unaffected, matching the standard `Initial` semantics.

Reviewer 2: <img width="225" height="169" alt="Image" src="https://github.com/user-attachments/assets/926efd6d-0982-4ea7-8028-dfd03135c53a" />

Reviewer 2: Are you using an LLM? I've asked for this list to be removed twice, but it's still not removed.
Policy is to disclose LLM use.

Author: My bad, using an LLM to help with the AEP text occasionally but reviewing everything, might have missed it. AFK right now, but let me check the policy for LLM when I look into this.

Author: Sorry but not sure how I missed removing this list, anyway its removed now. Thanks for pointing it out, will be more careful from now

- `Initial` → the admission-controller-side injection is now eligible to run on subsequent new-pod creations. Pods created during the window (or pods that pre-dated the VPA) are unaffected, matching the standard `Initial` semantics.

> - `Off` → `initialDelaySeconds` has no effect; `Off` already suppresses actuation indefinitely.

Reviewer 2: This is just describing existing features, no need to list it here

Author: Sure, makes sense, done

CPU Startup Boost (AEP-7862) still applies during the delay window: the boost is a startup-safety mechanism computed from the pod's own spec, not a recommendation, and the Admission Controller already applies it even under `updateMode: Off` (the `Off` short-circuit in `resource/vpa/matcher.go` is bypassed for VPAs with a boost). The gate reuses the same `Off` semantics, so boosted containers remain boosted.

The interaction appears at boost expiry, when the Updater performs the post-boost in-place downscale:

- Delay window still active → the pod is scaled down to its original spec resources; no recommendation is actuated.
- Delay window elapsed → the pod is scaled down to the VPA recommendation, per normal post-boost behaviour.

### Validation

Enforced via CRD schema:

- `initialDelaySeconds` must be an integer between `1` and `7776000` (90 days) if set. The upper bound follows the [API conventions for numeric fields](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#numeric-fields); observation windows are realistically hours to days, so 90 days is generous headroom while still rejecting absurd values.

> No dynamic (admission-webhook) validation is required in v1. All lifecycle transitions are handled declaratively by the gate.

Reviewer 2: I think it's preferable to also include validation in the webhook, since it can return much nicer error messages

Author: Agreed — added webhook validation alongside the CRD schema; same range check in resource/vpa/validation.go with descriptive errors

### Status Condition

Add a new value to the existing `VerticalPodAutoscalerStatus.Conditions` slice:

- **Type:** `InitialDelayActive`
- **Status:** `True` while the gate is active, `False` once it has elapsed.
- **Reason:** `WindowActive` (True) / `WindowExpired` (False).
- **Message:** human-readable summary including expiry timestamp.

> This lets `kubectl describe vpa` surface the gate without an operator having to compute `CreationTimestamp + initialDelaySeconds` mentally.

Reviewer 2: This is technically part of the API, I think it should be mentioned there too

Author: Sure, done 👍

- `admission-controller` — gates acceptance of the field on new/updated VPAs behind the feature gate, and honours the gate by skipping recommendation injection for pods created during an active window.

Enabling the feature gate causes:

- `admission-controller` to accept `initialDelaySeconds` on new/updated VPAs, and to skip recommendation injection for pods created while a VPA's window is active.
- `updater` to honour the gate.

Disabling the feature gate causes:

- `admission-controller` to reject new VPAs that set `initialDelaySeconds`, with a descriptive error message indicating the feature gate is disabled.
- `updater` and `admission-controller` to ignore the field on existing objects — the VPA behaves as if the field were not set (fail-open). This is deliberate: disabling the gate should never *increase* actuation restrictions on existing objects.

### Graduation Criteria

**Alpha (initial release):**

- Feature gate `VPAInitialDelay` disabled by default.
- Unit and e2e coverage as described in the [Test Plan](#test-plan).

**Alpha → Beta (gate enabled by default):**

- E2e tests stable for at least one release.
- No open bugs against the feature gate.

> - Positive user feedback on the field semantics (in particular the `CreationTimestamp`-anchored PATCH behaviour).

Reviewer 2: I'm not sure I understand what "the `CreationTimestamp`-anchored PATCH behaviour" means
May be remove everything in the braces?

Author: Done 👍

> **Beta → GA (gate locked on, then removed):**

Reviewer 2: ```suggestion
**Beta → GA (gate locked to enabled):**
```

We remove it in the following release, but no need to mention that here

> - No bug reports attributable to the feature for two consecutive releases.

Reviewer 2: We normally promote to GA in the release following beta, no need to wait for 2 consecutive releases

Author: Done

- **Old Updater, new Admission Controller** — pods created during the window are admitted without injection, but the Updater may still evict or resize during the window.
- **New Updater, old Admission Controller** — no evictions or resizes during the window, but pods created during the window still receive recommendation injection.

Neither skew causes errors or corrupted state; the failure mode is only that part of the gate is not honoured until the rollout completes. Operators who need strict gating should enable the feature gate only after all components are upgraded.

### Kubernetes Version Compatibility

This feature is entirely internal to the VPA controllers and depends on no new Kubernetes APIs. It is compatible with any Kubernetes version supported by the corresponding VPA release.

## Test Plan

> **Unit tests** (`pkg/updater/logic/`):

Reviewer: ```suggestion
**Updater tests:**
```

- `inInitialDelayWindow` returns `false` when `initialDelaySeconds` is `nil`, `0`, or absent.
- Gate short-circuits the eligibility path identically to `updateMode: Off` for `Recreate`, `InPlaceOrRecreate`, `InPlace`, `Initial`.
- Gate has no effect when `updateMode: Off` — behavior matches plain `updateMode: Off` regardless of the field's value.
- Metric and condition are set correctly on both sides of the transition.
- PATCH scenarios: shortening releases the gate, extending stays gated, removing closes the gate, extending after expiry re-arms only when the new expiry still lies in the future.

**Admission-controller tests:**

- Accept `initialDelaySeconds >= 1` on all update modes, including `Off` (no-op semantics).
- Reject `initialDelaySeconds < 1` and `initialDelaySeconds > 7776000` via CRD schema validation.
- Reject `initialDelaySeconds` on any VPA when the `VPAInitialDelay` feature gate is disabled.
- Pods created while the gate is active are admitted with their original spec resources, for every update mode.
- Pods created after the window elapses receive recommendation injection per the configured mode's normal semantics.

**E2E tests** (`e2e/`):

> - Create a VPA with `updateMode: Recreate` and `initialDelaySeconds: 60`. Verify no eviction occurs during the first 60 s. Advance time past the window; verify the next reconcile evicts as expected.

Reviewer 2: We have a lot of tests that just wait for time to pass, and I'd like to change that.

Can this rather do something like:
1. Create a VPA with `updateMode: Recreate` and `initialDelaySeconds: 60`, and a recommendation high enough to expect eviction
2. Watch for `InitialDelayActive` to be set to `True` and no pods evicted
3. Modify VPA to set `initialDelaySeconds` to 1
4. Watch for `InitialDelayActive` to be set to `False` and pods are evicted

Author: Restructured the e2e plan around your pattern - condition watches + patching the delay, no time-based waits.

- Create a VPA with `updateMode: Recreate`, `initialDelaySeconds: 60`, and a recommendation far enough from the pod's requests that eviction is expected. Watch for `InitialDelayActive=True` and verify no pods are evicted. Patch `initialDelaySeconds` to `1`; watch for `InitialDelayActive=False` and verify pods are evicted.

> - Repeat for `InPlaceOrRecreate`, `InPlace`, and `Initial` (for `Initial`, assert on injection at new-pod creation instead of eviction).

Reviewer 2: Let's not add too many e2e tests, I think these can be removed.
In theory they should be covered in integration tests, which we don't currently have

I do have a local patch to add integration tests to the updater, I wonder if I should try push that in before this AEP

@omerap12 thoughts here?

Reviewer: Yeah, I agree. If we can avoid e2e tests, we should definitely do that. This AEP has pretty straightforward logic, so it makes sense.

> I wonder if I should try to push that in before this AEP

If you can get that in beforehand, that would be great!

Reviewer 2: Let's change the wording here to include integration tests, and I'll see if I can get something before this goes live.

Author: Done 👍

This feature moves the sequencing into the VPA API itself. Operators declare "wait N seconds after creation before touching pods" alongside the update mode, and the Updater honours it.

### Goals

- Provide a declarative per-VPA field that delays actuation of recommendations — evictions, in-place resizes, and admission-time injection into newly created pods — for a configurable duration after VPA creation. CPU Startup Boost is not delayed (see [Interaction with CPU Startup Boost](#interaction-with-cpu-startup-boost)).
- Keep the Recommender's behaviour unchanged: recommendations are still computed and published to `status.recommendation` during the window, so operators can inspect them.
- Work uniformly across `Recreate`, `InPlaceOrRecreate`, `InPlace`, and `Initial`.
- Surface the gate state through both a status condition (`InitialDelayActive`) and a Prometheus gauge (`vpa_initial_delay_active`), so operators and dashboards can see when a VPA is gated without inspecting `spec`.

### Non-Goals

> - **Recommender-confidence-based gating.** This feature is a user-declared policy, not a recommender-computed signal. Interoperation with `LowConfidence` / `RecommendationProvided` is left for future work.

Reviewer 3: I would not mention a mechanism or feature in the document that is not used in the current implementation of the VPA. If you search for the string "LowConfidence" in the codebase, you will find that it only exists in the API definition

Author: Oh, didn't know this, thanks, removed

- **Reset semantics on modification of other fields.** Changing `resourcePolicy`, `targetRef`, or the target workload does not reset the window. The window is anchored on `vpa.CreationTimestamp`; users who need to re-observe a workload after a material change should delete and recreate the VPA.
- **Cluster-wide default windows.** Out of scope for this AEP; operators can inject a default with a Mutating Admission Policy (see [Alternatives Considered](#alternatives-considered), item 3).

## Proposal

Add a single optional field to `PodUpdatePolicy` (autoscaling.k8s.io/v1):

```go
// InitialDelaySeconds is the duration in seconds after this VPA's
// creation during which recommendations are not actuated: the Updater
// performs no evictions or in-place resizes, and the Admission
// Controller does not apply recommendations to newly created pods,
// regardless of the configured UpdateMode.
// +optional
// +kubebuilder:validation:Minimum=1
// +kubebuilder:validation:Maximum=7776000

> InitialDelaySeconds *int32 `json:"initialDelaySeconds,omitempty"`

Reviewer 3: If I put myself in the shoes of operators / users, I don't think using seconds as the unit is appropriate, since I assume users will typically want to configure values in the range of minutes or hours. Requiring them to convert minutes, hours into seconds adds unnecessary mental overhead. So let me ask: would adding a field that accepts duration values (for example, 10m or 24h) be more suitable in this case?

```go
InitialDelay *metav1.Duration `json:"initialDelay,omitempty"`
```

Reviewer 2: The API convention recommends against this: https://github.com/kubernetes/community/blob/main/contributors/devel/sig-architecture/api-conventions.md#units

Reviewer 3: hmm interesting, good to know!

2. The Recommender begins computing recommendations on the normal schedule and populates `status.recommendation`. It is unaware of and unaffected by the observation window.
3. On every Updater reconcile, before deciding whether the VPA is eligible for actuation, the Updater evaluates the gate. If the gate is active, the VPA is treated as if `updateMode` were `Off` for that reconcile: no pods are evicted (`Recreate` / `InPlaceOrRecreate`) and no in-place resize is attempted (`InPlace` / `InPlaceOrRecreate`).
4. The Admission Controller evaluates the same gate whenever a pod matching the VPA's target is created. While the gate is active it does not patch the pod's resources — the pod is admitted with its original spec, exactly as under `updateMode: Off`. This applies to every mode, not just `Initial`: without it, pods created during the window (scale-ups, node replacements, crash restarts) would receive un-stabilised recommendations at admission time.
5. The Updater sets the `InitialDelayActive` status condition to `True` while the gate is active and `False` once it has elapsed. It emits the `vpa_initial_delay_active` gauge accordingly.

> 6. On the first reconcile after `CreationTimestamp + initialDelaySeconds`, the gate opens and the configured `updateMode` takes effect on subsequent reconciles and pod creations.

Reviewer 3: Does this sentence make sense (6)? I think the first part of the sentence should be changed to something like this:

After the window elapses (i.e. now >= CreationTimestamp + InitialDelaySeconds), the gate opens etc....

Author: Sure, your version reads simpler. reworded.

### Interaction with `updateMode`

The observation window is orthogonal to `updateMode`. While the gate is active, the Updater and the Admission Controller behave as if `updateMode` were `Off` regardless of the configured value: no evictions, no in-place resizes, and no recommendation injection into pods created during the window.

### Interaction with CPU Startup Boost

CPU Startup Boost (AEP-7862) still applies during the delay window: the boost is a startup-safety mechanism computed from the pod's own spec, not a recommendation, and the Admission Controller already applies it even under `updateMode: Off` (the `Off` short-circuit in `resource/vpa/matcher.go` is bypassed for VPAs with a boost). The gate reuses the same `Off` semantics, so boosted containers remain boosted.

The interaction appears at boost expiry, when the Updater performs the post-boost in-place downscale:

- Delay window still active → the pod is scaled down to its original spec resources; no recommendation is actuated.

> - Delay window elapsed → the pod is scaled down to the VPA recommendation, per normal post-boost behaviour.

Reviewer 3: I think there will be cases where, after the window has elapsed, the pod's resource requests and limits are increased compared to their values during startupBoost, not just decreased. So I wouldn't use the term "scaled down" here.

Author: Yes, you're correct, updated this

### Interaction with Checkpoints

`VerticalPodAutoscalerCheckpoint` objects persist the Recommender's learned usage history for each VPA. When a VPA is deleted, its checkpoints are garbage-collected on the Recommender's next pass (`--checkpoints-gc-interval`, 10 minutes by default) — so if a VPA is recreated with the same name before that pass, the Recommender restores its history from the checkpoint and `status.recommendation` is immediately repopulated with mature values, rather than starting cold.

This matters here because a recreated VPA can therefore hold mature recommendations while its delay window is still active, which raises the question of whether the gate should open early in that case.

> Proposed handling: the gate ignores checkpoints entirely. A recreated VPA always starts a fresh window anchored to the new object's `CreationTimestamp`, whether its recommendations were restored or rebuilt from scratch.

Reviewer: We need to make sure this is documented extremely well. It's a bit counterintuitive how the VPA works (IMO) - it uses the checkpoint object so that the VPA object isn't treated as "new."

Author: Understood, have updated it to try to explain it better

> Rationale: keeping the gate a pure function of `CreationTimestamp` and the spec value is the core simplicity of this design — making it checkpoint-aware would reintroduce state and turn the field into a confidence signal, which is an explicit [non-goal](#non-goals). Users who recreate a VPA and are satisfied with the restored recommendations already have an escape hatch: shorten or remove `initialDelaySeconds` on the new object to open the gate early.

Reviewer 3: I like this approach, we don't need to complicate things. At the same time, if we roll out this feature and it reaches beta, we won't be able to change the behavior of the InitialDelaySeconds field without introducing a new API version (for example, to make it checkpoint object aware).

### Validation

`initialDelaySeconds` must be an integer between `1` and `7776000` (90 days) if set. The upper bound follows the [API conventions for numeric fields](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#numeric-fields); observation windows are realistically hours to days, so 90 days is generous headroom while still rejecting absurd values.

The range is enforced in two places:

- **CRD schema** — `Minimum=1`, `Maximum=7776000`, as the backstop.
- **Admission webhook** — the same check added to the existing VPA validation in `pkg/admission-controller/resource/vpa/validation.go`, which returns descriptive error messages, and which is also where the field is rejected when the `VPAInitialDelay` feature gate is disabled.

No cross-field validation or restrictions on modifying the field are required: all lifecycle transitions are handled declaratively by the gate.

### Status Condition

> Add a new value to the existing `VerticalPodAutoscalerStatus.Conditions` slice:

Reviewer 3: are you proposing that the updater sets this condition and that it is used only by VPA users to know more easily if the window is elapsed or not? The updater and the admission controller would still calculate the window independently all the time, right?

Author: Exactly right. The updater sets InitialDelayActive purely as a status output for visibility (kubectl describe, dashboards). Neither controller reads it, the updater and admission-controller each evaluate the gate independently on every pass from CreationTimestamp + initialDelaySeconds.

- **Reason:** `WindowActive` (True) / `WindowExpired` (False).
- **Message:** human-readable summary including expiry timestamp.

This lets `kubectl describe vpa` surface the gate without an operator having to compute `CreationTimestamp + initialDelaySeconds` mentally.

### Metric

Emit a new gauge from the Updater:

```
vpa_initial_delay_active{namespace, name} = 0 | 1
```

Value is `1` while the gate is active for that VPA, `0` otherwise. Enables fleet-wide visibility into how many VPAs are currently gated, useful for dashboards and alerting on stuck windows.

### Feature Enablement and Rollback

Feature gate: **`VPAInitialDelay`**.

Components consuming the gate:

- `updater` — reads `initialDelaySeconds` and honours the gate.
- `admission-controller` — gates acceptance of the field on new/updated VPAs behind the feature gate, and honours the gate by skipping recommendation injection for pods created during an active window.

Enabling the feature gate causes:

- `admission-controller` to accept `initialDelaySeconds` on new/updated VPAs, and to skip recommendation injection for pods created while a VPA's window is active.

> - `updater` to honour the gate.

Reviewer 3: nit: this part seems a little off to me, it does not feel entirely consistent. what we need to emphasize here is that both the updater and the admission controller read InitialDelaySeconds, and both honor the feature gate.

Author: Good nit, I think it a little asymmetric. Cleaned it up a bit.

**4. Do nothing; rely on operator tooling.** The status quo. Ships the complexity downstream to every VPA operator; the CronJob-plus-patch-controller dance ends up reinvented per environment.

**5. Immutable field after creation.** Reject PATCH that modifies `initialDelaySeconds` via admission. Simpler in that it eliminates the re-arm and shortening cases, but forces users into delete-and-recreate to change the window — an expensive operation that also discards the VPA's recommendation history. The proposed stateless-gate design gets the same "one-line mental model" without that cost.

**6. Reset on "material" spec changes** (`resourcePolicy`, `targetRef`). Arguably more correct — a workload with a different target really does have different history semantics — but forces us to define "material" and requires a status field to track the current anchor. Deferred: this AEP does not preclude adding such semantics in a future revision, and the `CreationTimestamp` anchor is a strict subset of any future anchor policy.

> **7. Reject PATCH that re-arms an expired gate.** Split the difference on (5) and the proposed design: allow PATCH freely during the window, but reject PATCHes that would push `expiry` past `now` once the window has already elapsed. Rejected because it makes the admission logic dependent on real time (surprising) and does not eliminate the underlying "PATCHes can change gate state" surprise — the `InitialDelayActive` condition and gauge already surface any re-arm immediately.

Reviewer: I don't understand this. It feels heavily AI-generated and doesn't follow a clear logical flow.

Let's remove everything under the `Alternatives Considered` section except points 1, 3, and 4.

Author: Sure, but 5, 6 and 7 were more about how to implement the observation window feature and what type of patches are allowed. 

So for example, explaining (7), rather than the current design (users can modify the initialDelaySeconds as they wish) is about disallowing users if they modify the initialDelaySeconds that it re-arms the expired gate. 
or in (6), ideally the gate should reset when something significant changes, for example, target deployment or resource policy or VPA mode, but has the added complication of deciding what is "significant" and what is not. 

Maybe I can re-word it so it doesn't seem AI generated? Or is the "Alternatives Considered" section more about alternatives to this solution (observation window) and not about alternative approaches for implementing this? Or maybe I can also collapse all 5, 6 and 7 into a single point.

Reviewer: The **Alternatives** section (I'm not sure why it was renamed to **Alternatives Considered**) should describe how users could achieve the **Per-VPA Observation Window** functionality using the existing VPA and native Kubernetes capabilities, without the proposed changes.

For example, how does item (6) address this requirement?

Reviewer: The all section wording feels extremely AI generated. so please let's remove all of that except the above.

Author: Reworked this section, thanks for the feedback

    // Updater and the Admission Controller treat the VPA as if
    // UpdateMode were Off: no evictions, no in-place resizes, and no
    // recommendation injection into newly created pods. After the window
    // elapses (now >= CreationTimestamp + InitialDelaySeconds), the
    // configured UpdateMode takes effect on subsequent reconciles and
    // pod creations.
    //
    // The window is computed as a pure function of vpa.CreationTimestamp
    // and this spec field. The field may be modified at any time; the
    // new value takes effect on the next reconcile. Removing the field or setting it to zero
    // closes the gate immediately. Setting a larger value after the
    // window has expired re-arms the gate only if the new expiry
    // (CreationTimestamp + InitialDelaySeconds) still lies in the
    // future; otherwise the gate stays open.
    //
    // If UpdateMode is Off, this field has no effect: Off already
    // suppresses Updater actuation indefinitely.
    //
    // Must be between 1 and 7776000 (90 days) if set.
    // +optional
    // +kubebuilder:validation:Minimum=1
    // +kubebuilder:validation:Maximum=7776000

>     InitialDelaySeconds *int32 `json:"initialDelaySeconds,omitempty"`

Reviewer: nit: this feels a bit excessive to me. We don't need to explain every feature in the comments. Since this is just an AEP, it's fine for now, but for the implementation I'd recommend trimming it down.

Reviewer 3: I agree with Omer. Furthermore I'm also struggling to understand the first part:

```
    // InitialDelaySeconds is the duration in seconds after this
    // VPA's creation during which recommendations are not actuated,
    // regardless of the configured UpdateMode.
```

IMO this is a better alternative:

```
// InitialDelaySeconds specifies the number of seconds to wait after
// the VPA object is created before recommendations are actuated,
// regardless of the configured UpdateMode.
```

Author: Sure makes sense, I have updated this section and trimmed the entire comment

**2. Recommender-side confidence gate.** Block `RecommendationProvided` from becoming `True` for N hours or until a confidence threshold is met. Cleaner conceptually but hides visibility — operators would see no recommendation at all in `status.recommendation` during the window, defeating the "collect first, inspect, then apply" workflow this feature is meant to support.

**3. Global default flag** (e.g. `--default-initial-delay-seconds`). A sane cluster-wide default is genuinely useful, but the flag would need to be added to — and kept in sync between — both the Updater and the Admission Controller, since both evaluate the gate. Operators who want a cluster-wide default can instead inject one at admission time with a [Mutating Admission Policy](https://kubernetes.io/docs/reference/access-authn-authz/mutating-admission-policy/), with no new VPA configuration surface. For example, to default every VPA that does not set the field to a one-hour window:

```yaml

> apiVersion: admissionregistration.k8s.io/v1beta1

Reviewer: We don't need to provide a solution for a problem we're not going to create, right?

I think it's enough to state that a single global flag for the updater and the admission controller isn't sufficient. I'm not sure why the recommender would need this flag, so calling it "cluster-wide" doesn't seem accurate. 

I'd also remove all the excessive detail about the mutating admission policy and related content.

Author: Sure, made it more compact
